src/arcade/patch/agent/action/PatchActionTreat.java

```diff
@@ -0,0 +1,378 @@
+            findLocations(comp, type, locs, siteLocs, sim);
+            Utilities.shuffleList(siteLocs, sim.random);
+            // sort locations in descending order from highest to lowest density
+            siteLocs.sort(Comparator.comparingInt(l -> -computeDensity(grid, l)));
+            insert(siteLocs, simstate, pop);
+        }
+    }
+
+    /**
+     * Helper method to find possible locations to insert T-cells.
+     *
+     * @param comp the component
+     * @param type the type of component (source, pattern, or graph)
+     * @param locs the locations to check
+     * @param siteLocs the locations that meet the criteria
+     * @param sim the simulation instance
+     * @throws IllegalArgumentException if the component type is invalid
+     */
+    private void findLocations(
+            PatchComponentSites comp,
+            String type,
+            ArrayList<LocationContainer> locs,
+            ArrayList<Location> siteLocs,
+            PatchSimulation sim) {
+        if (type.equals("graph")) {
+            findGraphSites(comp, locs, siteLocs);
+        } else if (type.equals("source") || type.equals("pattern")) {
+            double[][][] damage;
+            boolean[][][] sitesLat;
+
+            if (type.equals("source")) {
+                damage = ((PatchComponentSitesSource) comp).getDamage();
+                sitesLat = ((PatchComponentSitesSource) comp).getSources();
+            } else {
+                damage = ((PatchComponentSitesPattern) comp).getDamage();
+                sitesLat = ((PatchComponentSitesPattern) comp).getPatterns();
+            }
+            pruneSite(locs, sim, damage, sitesLat, siteLocs);
+        } else {
+            throw new IllegalArgumentException(
+                    "Invalid component type: "
+                            + type
+                            + ". Must be of type source, pattern, or graph.");
+        }
+    }
+
+    /**
+     * Helper method to check if radius is wide enough for T-cells to pass through.
+     *
+     * @param comp the component
+     * @param locs the locations to check
+     * @param siteLocs the locations that meet the criteria
+     */
+    private void findGraphSites(
+            PatchComponentSites comp,
+            ArrayList<LocationContainer> locs,
+            ArrayList<Location> siteLocs) {
+        Graph graph = ((PatchComponentSitesGraph) comp).getGraph();
+        Bag allEdges = new Bag(graph.getAllEdges());
+        PatchComponentSitesGraph graphSites = (PatchComponentSitesGraph) comp;
+
+        Set<Coordinate> coordinateSet =
+                locs.stream()
+                        .map(container -> ((PatchLocationContainer) container).coordinate)
+                        .collect(Collectors.toSet());
+
+        for (Object edgeObj : allEdges) {
+            SiteEdge edge = (SiteEdge) edgeObj;
+            Bag allEdgeLocs = new Bag();
+            if (Objects.equals(coord, "Hex")) {
```

Comment:
why not just graphSites.getSpan() rather than casting and specifying the type of the graphsites?